internal/extender/unschedulablepods.go

```diff
@@ -0,0 +1,135 @@
+// limitations under the License.
+
+package extender
+
+import (
+	"context"
+	"time"
+
+	"github.com/palantir/k8s-spark-scheduler-lib/pkg/resources"
+	"github.com/palantir/witchcraft-go-logging/wlog/svclog/svc1log"
+	"github.com/palantir/witchcraft-go-logging/wlog/wapp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+)
+
+const (
+	podExceedsClusterCapacity v1.PodConditionType = "PodExceedsClusterCapacity"
+)
+
+// UnschedulablePodMarker checks for spark scheduler managed pending driver pods
+// and checks if they can fit if the cluster was empty, else marks them with a
+// custom pod condition.
+type UnschedulablePodMarker struct {
+	nodeLister       corelisters.NodeLister
+	podLister        corelisters.PodLister
+	coreClient       corev1.CoreV1Interface
+	overheadComputer *OverheadComputer
+	binpacker        *Binpacker
+}
+
+// NewUnschedulablePodMarker creates a new UnschedulablePodMarker
+func NewUnschedulablePodMarker(
+	nodeLister corelisters.NodeLister,
+	podLister corelisters.PodLister,
+	coreClient corev1.CoreV1Interface,
+	overheadComputer *OverheadComputer,
+	binpacker *Binpacker) *UnschedulablePodMarker {
+	return &UnschedulablePodMarker{
+		nodeLister:       nodeLister,
+		podLister:        podLister,
+		coreClient:       coreClient,
+		overheadComputer: overheadComputer,
+		binpacker:        binpacker,
+	}
+}
+
+// Start starts periodic scanning for unschedulable pods
+func (u *UnschedulablePodMarker) Start(ctx context.Context) {
+	_ = wapp.RunWithFatalLogging(ctx, u.doStart)
+}
+
+func (u *UnschedulablePodMarker) doStart(ctx context.Context) error {
+	t := time.NewTicker(60 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-t.C:
+			u.mark(ctx)
+		}
+	}
+}
+
+func (u *UnschedulablePodMarker) mark(ctx context.Context) {
+	pods, err := u.podLister.List(labels.Everything())
+	if err != nil {
+		svc1log.FromContext(ctx).Error("failed to list pods", svc1log.Stacktrace(err))
+		return
+	}
+	now := time.Now()
+	for _, pod := range pods {
+		if pod.Spec.SchedulerName == SparkSchedulerName && // TODO: constant
+			len(pod.Spec.NodeName) == 0 &&
+			pod.DeletionTimestamp == nil &&
+			pod.Labels[SparkRoleLabel] == Driver &&
+			pod.CreationTimestamp.Time.Add(10*time.Minute).Before(now) {
+			err := u.markPod(ctx, pod)
+			if err != nil {
+				svc1log.FromContext(ctx).Error("failed to check if pod was unschedulable",
+					svc1log.SafeParam("podName", pod.Name),
+					svc1log.SafeParam("podNamespace", pod.Namespace),
+					svc1log.Stacktrace(err))
+			}
+		}
+
+	}
+}
+
+func (u *UnschedulablePodMarker) markPod(ctx context.Context, driver *v1.Pod) error {
+	nodes, err := u.nodeLister.List(labels.Set(driver.Spec.NodeSelector).AsSelector())
+	if err != nil {
+		return err
+	}
+	nodeNames := make([]string, 0, len(nodes))
```

Comment:
nit - I'm guessing all this boilerplate for getting a list of node names is copied from elsewhere, can we move to a utility function?

Reply:
I suppose there should be a function `resources.availableResources()` that hides the node listing + overhead computation

Reply:
they do slightly different things, we don't add resource reservation usage here